### PR TITLE
feat(dynatrace): handle non-gauge data types

### PR DESCRIPTION
**Pull request by dyladan**
Part 3 of splitting up #4853 into smaller PRs

Previously all metrics were treated as if they were a gauge. This introduces proper handling for counters and histograms.

### CODE REVIEW

**Comment by dyladan** [comment]
The test failure was in the k8sclusterreceiver not the package I modified. Can someone with permission please re-run the test suite?

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 1]
```diff
@@ -109,19 +121,16 @@ func (e *exporter) serializeMetrics(md pdata.Metrics) []string {
...
-					l = serialization.SerializeHistogramMetrics(e.cfg.Prefix, metric.Name(), metric.Histogram().DataPoints(), e.defaultDimensions)
+				metricLines, err := serialization.SerializeMetric(e.cfg.Prefix, metric, e.defaultDimensions, e.staticDimensions, e.prevPts)
+
+				if err != nil {
+					e.logger.Sugar().Errorf("failed to serialize metric %s", err.Error())
+				}
+
+				if len(metricLines) > 0 {
+					lines = append(lines, metricLines...)
+					e.logger.Debug(fmt.Sprintf("Exporting type %s, Name: %s, len: %d ", metric.DataType().String(), metric.Name(), len(metricLines)))
```
Comment (arminru) [inline exporter/dynatraceexporter/metrics_exporter.go:132]:
```suggestion
					e.logger.Debug(fmt.Sprintf("Serialized type %s, Name: %s, len: %d ", metric.DataType().String(), metric.Name(), len(metricLines)))
```
The actual export (what I'd read as already sending it) comes later actually.

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 2]
```diff
@@ -109,19 +121,16 @@ func (e *exporter) serializeMetrics(md pdata.Metrics) []string {
...
-				case pdata.MetricDataTypeGauge:
-					l = serialization.SerializeNumberDataPoints(e.cfg.Prefix, metric.Name(), metric.Gauge().DataPoints(), e.defaultDimensions)
-				case pdata.MetricDataTypeSum:
-					l = serialization.SerializeNumberDataPoints(e.cfg.Prefix, metric.Name(), metric.Sum().DataPoints(), e.defaultDimensions)
-				case pdata.MetricDataTypeHistogram:
-					l = serialization.SerializeHistogramMetrics(e.cfg.Prefix, metric.Name(), metric.Histogram().DataPoints(), e.defaultDimensions)
+				metricLines, err := serialization.SerializeMetric(e.cfg.Prefix, metric, e.defaultDimensions, e.staticDimensions, e.prevPts)
+
+				if err != nil {
+					e.logger.Sugar().Errorf("failed to serialize metric %s", err.Error())
```
Comment (arminru) [inline exporter/dynatraceexporter/metrics_exporter.go:127]: You could add the metric type and name like below if that's not always included in the error message.

File: exporter/dynatraceexporter/serialization/serialization.go [hunk 3]
```diff
@@ -15,100 +15,313 @@
...
-		dimsFromMetric = append(dimsFromMetric, dimensions.NewDimension(k, v.AsString()))
-		return true
-	})
-	return dimensions.NewNormalizedDimensionList(dimsFromMetric...)
-}
+func SerializeMetric(prefix string, metric pdata.Metric, defaultDimensions, staticDimensions dimensions.NormalizedDimensionList, prev *ttlmap.TTLMap) ([]string, error) {
+	var metricLines []string
+	switch metric.DataType() {
+	case pdata.MetricDataTypeNone:
+		return nil, nil
```
Comment (arminru) [inline exporter/dynatraceexporter/serialization/serialization.go:32]: Would it make sense to log this (by returning an err)?

**Comment by dyladan** [comment]
The broken lint is not in the dynatrace package so I'm not sure what the deal is there.

**Review by arminru — APPROVED** [review]

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 4]
```diff
@@ -109,19 +121,16 @@ func (e *exporter) serializeMetrics(md pdata.Metrics) []string {
...
-				case pdata.MetricDataTypeGauge:
-					l = serialization.SerializeNumberDataPoints(e.cfg.Prefix, metric.Name(), metric.Gauge().DataPoints(), e.defaultDimensions)
-				case pdata.MetricDataTypeSum:
-					l = serialization.SerializeNumberDataPoints(e.cfg.Prefix, metric.Name(), metric.Sum().DataPoints(), e.defaultDimensions)
-				case pdata.MetricDataTypeHistogram:
-					l = serialization.SerializeHistogramMetrics(e.cfg.Prefix, metric.Name(), metric.Histogram().DataPoints(), e.defaultDimensions)
+				metricLines, err := serialization.SerializeMetric(e.cfg.Prefix, metric, e.defaultDimensions, e.staticDimensions, e.prevPts)
+
+				if err != nil {
+					e.logger.Sugar().Errorf("failed to serialize %s %s, %s", metric.DataType().String(), metric.Name(), err.Error())
```
Comment (arminru) [inline exporter/dynatraceexporter/metrics_exporter.go:127]:
```suggestion
					e.logger.Sugar().Errorf("Failed to serialize %s '%s': %s", metric.DataType().String(), metric.Name(), err.Error())
```

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 5]
```diff
@@ -109,19 +121,16 @@ func (e *exporter) serializeMetrics(md pdata.Metrics) []string {
...
+				}
+
+				if len(metricLines) > 0 {
+					lines = append(lines, metricLines...)
+					e.logger.Debug(fmt.Sprintf("Serialized %s %s - %d lines", metric.DataType().String(), metric.Name(), len(metricLines)))
```
Comment (arminru) [inline exporter/dynatraceexporter/metrics_exporter.go:132]:
```suggestion
					e.logger.Debug(fmt.Sprintf("Serialized %s '%s' - %d lines", metric.DataType().String(), metric.Name(), len(metricLines)))
```

**Review by arminru — APPROVED** [review]

File: exporter/dynatraceexporter/serialization/histogram.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	if t == pdata.AggregationTemporalityCumulative {
+		// convert to delta histogram
+		// skip first point because there is nothing to calculate a delta from
+		// what if bucket bounds change
+		// TTL for cumulative histograms
+		// reset detection? if cumulative and count decreases, the process probably reset
+		return "", errors.New("cumulative histograms not supported")
+	}
+
+	min, max, nonEmpty := histMinMax(dp.ExplicitBounds(), dp.BucketCounts())
```
Comment (arminru) [inline exporter/dynatraceexporter/serialization/histogram.go]: You could check `dp.Count()` before and then also spare the `nonEmpty` flag.

File: exporter/dynatraceexporter/serialization/histogram.go [hunk 7]
```diff
@@ -32,9 +32,9 @@ func serializeHistogram(name, prefix string, dims dimensions.NormalizedDimension
 		return "", errors.New("cumulative histograms not supported")
 	}
 
-	min, max, nonEmpty := histMinMax(dp.ExplicitBounds(), dp.BucketCounts())
+	min, max := estimateHistMinMax(dp)
 
-	if !nonEmpty {
+	if dp.Count() == 0 {
```
Comment (arminru) [inline exporter/dynatraceexporter/serialization/histogram.go]: That could be put the other way round.

**Review by arminru — APPROVED** [review]

File: exporter/dynatraceexporter/metrics_exporter.go (hunk 5 above)
Comment (dyladan) [reply exporter/dynatraceexporter/metrics_exporter.go:132]: Do the quotes add anything here? Name will contain no spaces and neither with the type so it is unambiguous

File: exporter/dynatraceexporter/serialization/histogram.go (hunk 7 above)
Comment (dyladan) [reply exporter/dynatraceexporter/serialization/histogram.go]: Done

File: exporter/dynatraceexporter/metrics_exporter.go (hunk 4 above)
Comment (dyladan) [reply exporter/dynatraceexporter/metrics_exporter.go:127]: Seems like idiomatic go style is to keep lowercase for error strings
image

File: exporter/dynatraceexporter/metrics_exporter.go (hunk 5 above)
Comment (arminru) [reply exporter/dynatraceexporter/metrics_exporter.go:132]: Ah I wasn't sure about the type at this point but yeah then it's fine as-is, just wanted to make it consistent with the one above.

File: exporter/dynatraceexporter/metrics_exporter.go (hunk 4 above)
Comment (arminru) [reply exporter/dynatraceexporter/metrics_exporter.go]: I see, I just noticed it's inconsistent with the log in line 132 then.

**Comment by dyladan** [comment]
Looks like the loadtest portion of the build-publish step is failing? I don't think my PR could have caused that. Is the dynatrace exporter even included in the load test?

**Comment by dyladan** [comment]
Again the broken loadtest seems unrelated to our changes.

**Comment by bogdandrutu** [comment]
@dyladan ready to merge this?

**Comment by dyladan** [comment]
Yes

**Comment by dyladan** [comment]
@bogdandrutu i'm not sure whats going on with the failures here. the error messages aren't helpful but I don't think it's an issue in the dynatrace exporter.

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**